// firebase-tools/lib/deploy/functions/runtimes/node/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Delegate = exports.tryCreateDelegate = void 0;
const util_1 = require("util");
const fs = require("fs");
const path = require("path");
const error_1 = require("../../../../error");
const parseRuntimeAndValidateSDK_1 = require("./parseRuntimeAndValidateSDK");
const getProjectId = require("../../../../getProjectId");
const validate = require("./validate");
const logger_1 = require("../../../../logger");
const versioning = require("./versioning");
const parseTriggers = require("./parseTriggers");
async function tryCreateDelegate(context, options) {
    const sourceDirName = options.config.get("functions.source");
    const sourceDir = options.config.path(sourceDirName);
    const packageJsonPath = path.join(sourceDir, "package.json");
    if (!(await util_1.promisify(fs.exists)(packageJsonPath))) {
        logger_1.logger.debug("Customer code is not Node");
        return undefined;
    }
    let runtime = options.config.get("functions.runtime") || "";
    runtime = parseRuntimeAndValidateSDK_1.getRuntimeChoice(sourceDir, runtime);
    if (!runtime.startsWith("nodejs")) {
        logger_1.logger.debug("Customer has a package.json but did not get a nodejs runtime. This should not happen");
        throw new error_1.FirebaseError(`Unexpected runtime ${runtime}`);
    }
    return new Delegate(getProjectId(options), options.config.projectDir, sourceDirName, sourceDir, runtime);
}
exports.tryCreateDelegate = tryCreateDelegate;
class Delegate {
    constructor(projectId, projectDir, sourceDirName, sourceDir, runtime) {
        this.projectId = projectId;
        this.projectDir = projectDir;
        this.sourceDirName = sourceDirName;
        this.sourceDir = sourceDir;
        this.runtime = runtime;
        this.name = "nodejs";
        this._sdkVersion = "";
    }
    get sdkVersion() {
        if (!this._sdkVersion) {
            this._sdkVersion = versioning.getFunctionsSDKVersion(this.sourceDir) || "";
        }
        return this._sdkVersion;
    }
    validate() {
        versioning.checkFunctionsSDKVersion(this.sdkVersion);
        validate.packageJsonIsValid(this.sourceDirName, this.sourceDir, this.projectDir);
        return Promise.resolve();
    }
    async build() {
    }
    watch() {
        return Promise.resolve(() => Promise.resolve());
    }
    async discoverSpec(config, env) {
        return parseTriggers.discoverBackend(this.projectId, this.sourceDir, this.runtime, config, env);
    }
}
exports.Delegate = Delegate;
